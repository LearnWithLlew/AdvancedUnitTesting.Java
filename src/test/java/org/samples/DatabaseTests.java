package org.samples;


import com.spun.util.ObjectUtils;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.samples.CiUtils.skipTestOnCI;
import static org.samples.ProductionDatabase.createDatabase;

public class DatabaseTests {
    @Test
    void testConnection() throws SQLException {
        skipTestOnCI();
        assertEquals(42, SqlUtilites.executeNumericQuery("Select 42;", getConnection()));
    }

    @Test
    void testVersion() throws SQLException {
        skipTestOnCI();
        assertEquals(1, ProductionDatabase.getSchemaVersion(getConnection()));
    }
    @Test
    void testSchema() throws SQLException {
        skipTestOnCI();
        var schema = SqlUtilites.loadSchema(getConnection());
        Approvals.verify(SqlUtilites.print(schema));
    }

    @Test
    void testInMemoryConnection() throws SQLException {
        assertEquals(42, SqlUtilites.executeNumericQuery("Select 42;", getInMemoryConnection()));
    }
    @Test
    void testInMemoryVersion() throws SQLException {
        Connection connection = createInMemoryDatabase(1);
        assertEquals(1, ProductionDatabase.getSchemaVersion(connection));
    }

    
    void testInMemorySchema() throws SQLException {
        Connection connection = createInMemoryDatabase(1);
        var schema = SqlUtilites.loadSchema(connection);
        Approvals.verify(SqlUtilites.print(schema));
    }
    private Connection createInMemoryDatabase(int version) throws SQLException {
        Connection connection = getInMemoryConnection();
        createDatabase(connection, version);
        return connection;
    }

    private Connection getConnection() {
        try {
            return DriverManager.getConnection("jdbc:mariadb://localhost:3306/sakila", "root", "");
        } catch (SQLException e) {
            throw ObjectUtils.throwAsError(e);
        }
    }


    private Connection getInMemoryConnection() {
        try {
            Connection connection = DriverManager.getConnection("jdbc:h2:~/test", "sa", "");

            connection.createStatement().execute("DROP SCHEMA IF EXISTS sakila CASCADE");
            connection.createStatement().execute("CREATE SCHEMA sakila");
            connection.createStatement().execute("SET SCHEMA sakila");
            return connection;
        } catch (SQLException e) {
            throw ObjectUtils.throwAsError(e);
        }
    }
}
