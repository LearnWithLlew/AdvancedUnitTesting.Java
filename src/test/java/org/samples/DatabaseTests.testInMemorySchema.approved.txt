actor
--------------------
     actor_id: SMALLINT UNSIGNED[Nullable]
     first_name: VARCHAR[Nullable]
     last_name: VARCHAR[Nullable]
     last_update: TIMESTAMP[Nullable]


address
--------------------
     address_id: SMALLINT UNSIGNED[Nullable]
     address: VARCHAR[Nullable]
     address2: VARCHAR[Not Null]
     district: VARCHAR[Nullable]
     city_id: SMALLINT UNSIGNED[Nullable]
     postal_code: VARCHAR[Not Null]
     phone: VARCHAR[Nullable]
     last_update: TIMESTAMP[Nullable]


category
--------------------
     category_id: TINYINT UNSIGNED[Nullable]
     name: VARCHAR[Nullable]
     last_update: TIMESTAMP[Nullable]


city
--------------------
     city_id: SMALLINT UNSIGNED[Nullable]
     city: VARCHAR[Nullable]
     country_id: SMALLINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


country
--------------------
     country_id: SMALLINT UNSIGNED[Nullable]
     country: VARCHAR[Nullable]
     last_update: TIMESTAMP[Nullable]


customer
--------------------
     customer_id: SMALLINT UNSIGNED[Nullable]
     store_id: TINYINT UNSIGNED[Nullable]
     first_name: VARCHAR[Nullable]
     last_name: VARCHAR[Nullable]
     email: VARCHAR[Not Null]
     address_id: SMALLINT UNSIGNED[Nullable]
     active: BOOLEAN[Nullable]
     create_date: DATETIME[Nullable]
     last_update: TIMESTAMP[Not Null]


film
--------------------
     film_id: SMALLINT UNSIGNED[Nullable]
     title: VARCHAR[Nullable]
     description: TEXT[Not Null]
     release_year: YEAR[Not Null]
     language_id: TINYINT UNSIGNED[Nullable]
     original_language_id: TINYINT UNSIGNED[Not Null]
     rental_duration: TINYINT UNSIGNED[Nullable]
     rental_rate: DECIMAL[Nullable]
     length: SMALLINT UNSIGNED[Not Null]
     replacement_cost: DECIMAL[Nullable]
     rating: ENUM[Not Null]
     special_features: SET[Not Null]
     last_update: TIMESTAMP[Nullable]


film_actor
--------------------
     actor_id: SMALLINT UNSIGNED[Nullable]
     film_id: SMALLINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


film_category
--------------------
     film_id: SMALLINT UNSIGNED[Nullable]
     category_id: TINYINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


film_text
--------------------
     film_id: SMALLINT[Nullable]
     title: VARCHAR[Nullable]
     description: TEXT[Not Null]


inventory
--------------------
     inventory_id: MEDIUMINT UNSIGNED[Nullable]
     film_id: SMALLINT UNSIGNED[Nullable]
     store_id: TINYINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


language
--------------------
     language_id: TINYINT UNSIGNED[Nullable]
     name: CHAR[Nullable]
     last_update: TIMESTAMP[Nullable]


metadata
--------------------
     id: INT[Nullable]
     data_name: VARCHAR[Nullable]
     data_value: VARCHAR[Nullable]


payment
--------------------
     payment_id: SMALLINT UNSIGNED[Nullable]
     customer_id: SMALLINT UNSIGNED[Nullable]
     staff_id: TINYINT UNSIGNED[Nullable]
     rental_id: INT[Not Null]
     amount: DECIMAL[Nullable]
     payment_date: DATETIME[Nullable]
     last_update: TIMESTAMP[Not Null]


rental
--------------------
     rental_id: INT[Nullable]
     rental_date: DATETIME[Nullable]
     inventory_id: MEDIUMINT UNSIGNED[Nullable]
     customer_id: SMALLINT UNSIGNED[Nullable]
     return_date: DATETIME[Not Null]
     staff_id: TINYINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


staff
--------------------
     staff_id: TINYINT UNSIGNED[Nullable]
     first_name: VARCHAR[Nullable]
     last_name: VARCHAR[Nullable]
     address_id: SMALLINT UNSIGNED[Nullable]
     picture: BLOB[Not Null]
     email: VARCHAR[Not Null]
     store_id: TINYINT UNSIGNED[Nullable]
     active: BOOLEAN[Nullable]
     username: VARCHAR[Nullable]
     password: VARCHAR[Not Null]
     last_update: TIMESTAMP[Nullable]


store
--------------------
     store_id: TINYINT UNSIGNED[Nullable]
     manager_staff_id: TINYINT UNSIGNED[Nullable]
     address_id: SMALLINT UNSIGNED[Nullable]
     last_update: TIMESTAMP[Nullable]


